source/GameInterface/Services/MobileParties/Patches/UnitRecruitPatch.cs

```diff
@@ -0,0 +1,123 @@
+﻿using Common;
+using Common.Extensions;
+using Common.Logging;
+using Common.Messaging;
+using Common.Util;
+using GameInterface.Policies;
+using GameInterface.Services.MobileParties.Extensions;
+using GameInterface.Services.MobileParties.Handlers;
+using GameInterface.Services.MobileParties.Messages;
+using HarmonyLib;
+using Serilog;
+using Serilog.Core;
+using System;
+using System.Reflection;
+using TaleWorlds.CampaignSystem;
+using TaleWorlds.CampaignSystem.Actions;
+using TaleWorlds.CampaignSystem.Party;
+using TaleWorlds.CampaignSystem.Roster;
+using TaleWorlds.CampaignSystem.ViewModelCollection.CharacterDeveloper;
+
+namespace GameInterface.Services.MobileParties.Patches
+{
+    [HarmonyPatch(typeof(TroopRoster))]
+    public class UnitRecruitPatch
+    {
+        private static PropertyInfo TroopRoster_OwnerParty => typeof(TroopRoster).GetProperty("OwnerParty", BindingFlags.Instance | BindingFlags.NonPublic);
+
+        private static readonly MethodInfo TroopRoster_AddNewElement = typeof(TroopRoster).GetMethod("AddNewElement", BindingFlags.NonPublic | BindingFlags.Instance);
+
+        private static readonly ILogger Logger = LogManager.GetLogger<UnitRecruitPatch>();
+
+        [HarmonyPrefix]
+        [HarmonyPatch("AddNewElement")]
+        private static bool PrefixAddNewElement(TroopRoster __instance, CharacterObject character, bool insertAtFront = false, int insertionIndex = -1)
+        {
+            if (AllowedThread.IsThisThreadAllowed()) return true;
+
+            if (PolicyProvider.AllowOriginalCalls) return true;
+
+            PartyBase ownerParty = (PartyBase)TroopRoster_OwnerParty.GetValue(__instance);
+
+            if (ownerParty == null) return false;
+
+            MessageBroker.Instance.Publish(__instance, new NewTroopAdded(
+                character.StringId, 
+                ownerParty.MobileParty.StringId, 
+                __instance.IsPrisonRoster, 
+                insertAtFront, 
+                insertionIndex));
+
+            return false;
+        }
+
+        [HarmonyPrefix]
+        [HarmonyPatch(nameof(TroopRoster.AddToCountsAtIndex))]
+        private static bool PrefixAddToCountsAtIndex(TroopRoster __instance, int index, int countChange, int woundedCountChange = 0, int xpChange = 0, bool removeDepleted = true)
```

Comment:
I don't see how this patch on the client won't run with uncontrolled parties

Reply:
Currently, it does run with uncontrolled parties on both patches. The normal recruiting is only done on server (handled further up the call stack) but as fights are currently done locally only the client is aware of the changes in a battle thus have to change it. 

A possible middle-ground for now might be that you can only change the main party and any involved parties in the current map event of the main party? I'm not sure if there's a way to remedy this issue later as the battles will always be independent of the server.